Migrate test-integration-cli experimental ipvlan test to integration

vdemeester opened the pull request:
<del>Build on top of #36697, only the second commit (007b7f7) should be reviewed :angel: </del>

All `Ipvlan` related test on `DockerSuite` and `DockerNetworkSuite`
are migrated to `ipvlan_test.go`.

The end goal being to remove the `experimental` builds.

It also refactors macvlan tests a bit (quicker) *and* it put `macvlan` and `ipvlan` tests on there own folder in order to be isolated and have quick "suite).

cc @fcrisciani @selansen 

:lion: 

Signed-off-by: Vincent Demeester vincent@sbr.pm

vdemeester commented:
Updated to do a bit more than just `ipvlan` migration. It also refactors macvlan tests a bit (quicker) *and* it put `macvlan` and `ipvlan` tests on there own folder in order to be isolated and have quick "suite).

fcrisciani commented:
@ctelfer can you take a look too?

fcrisciani reviewed (approved):
high level I guess looks good

ctelfer reviewed (commented):
Sorry it has taken a while to get through the code.  LGTM overall.  Just one comment about a cleanup function.

vdemeester commented on integration/network/helpers.go:
```diff
@@ -0,0 +1,85 @@
+// DeleteInterface deletes a network interface
+func DeleteInterface(t *testing.T, ifName string) {
+	icmd.RunCommand("ip", "link", "delete", ifName).Assert(t, icmd.Success)
+	icmd.RunCommand("iptables", "-t", "nat", "--flush").Assert(t, icmd.Success)
```
@ctelfer oh interesting… I took what was previously here so the behavior doesn't change from previous tests. cc @mavenugo to get more insight :angel:

thaJeztah replied:
As discussed on slack: these tests set up a new daemon, so this shouldn't be a problem, but perhaps we can add a TODO to these (as a follow-up)

cpuguy83 replied:
The problem is these daemons aren't in their own network namespace and it's flushing everything in the main namespace.

thaJeztah commented on integration/network/ipvlan_test.go:
```diff
@@ -0,0 +1,543 @@
+
+func TestDockerNetworkIpvlanPersistance(t *testing.T) {
+	// verify the driver automatically provisions the 802.1q link (di-dummy0.70)
+	skip.If(t, testEnv.DaemonInfo.OSType != "linux")
```
Commented on another PR; we should go over the tests and be consistent (either "!= linux" or "== windows").

No need to fix for this PR, as it's a general thing to look at

thaJeztah commented on integration/network/macvlan_test.go:
```diff
@@ -189,7 +192,7 @@ func TestDockerNetworkMacvlanMultiSubnet(t *testing.T) {
 	skip.If(t, testEnv.IsRemoteDaemon())
 	skip.If(t, !macvlanKernelSupport(), "Kernel doesn't support macvlan")
-	t.Skip("Temporarily skipping while investigating sporadic v6 CI issues")
+	// t.Skip("Temporarily skipping while investigating sporadic v6 CI issues")
```
Guess this can be removed

thaJeztah commented on integration/network/macvlan_test.go:
```diff
@@ -48,292 +49,275 @@ func TestDockerNetworkMacvlanPersistance(t *testing.T) {
+		result, err := container.Exec(ctx, client, id1, []string{"ip", "route"})
+		assert.NilError(t, err)
+		assert.Check(t, strings.Contains(result.Combined(), "default via 172.28.130.1 dev eth0"))
+		// Validate macvlan bridge mode sets the v6 gateway to the user specified default gateway/next-hop
```
Wondering if we should start moving comments like this as custom message (not for this PR)

thaJeztah commented on integration/network/helpers.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+// CheckKernelMajorVersionGreaterOrEqualThen returns whether the kernel version is greater or equal than the one provided
+func CheckKernelMajorVersionGreaterOrEqualThen(kernelVersion int, majorVersion int) bool {
```
Should probably be moved somewhere else, as it's not network-specific (may even be another helper like this already)

thaJeztah reviewed (approved):
LGTM, left some nits/questions but it's ok to address those separately

cpuguy83 reviewed (approved):
LGTM

cpuguy83 commented:
And a conflict now.

vdemeester commented:
Gonna fix the conflicts and take @thaJeztah comments into account tomorrow morning (Paris time :wink:)

vdemeester committed:
Migrate test-integration-cli experimental ipvlan test to integration

All `Ipvlan` related test on `DockerSuite` and `DockerNetworkSuite`
are migrated to `ipvlan_test.go`.

The end goal being to remove the `experimental` builds.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Refactor macvlan tests a bit

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Move ipvlan and macvlan tests on their own folder…

… making each folder/suites quicker to run

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
Rebased :angel:
